docs/operators/deduper.md

```diff
@@ -0,0 +1,523 @@
+occurring more than once in the original data set.
+
+
+
+![](images/deduper/image00.png)
+
+
+Dedup - “How” in a Nutshell
+----------------------------------------
+
+In order to quickly decide whether an incoming tuple is duplicate
+or unique, it has to store each incoming tuple (or a signature, like key
+for example) to be used for comparison later. A plain in-memory storage
+may work for small datasets, but will not scale for large ones. Deduper employs a large scale distributed persistent hashing mechanism (known as the Managed State) which allows
+it to identify if a particular tuple is duplicate or unique.
+Each time it identifies a tuple as a unique tuple, it also
+stores it into the Managed state for future lookup.
+
+![](images/deduper/image04.png)
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+
+Following are the different components of the Deduper
+
+1.  **Dedup Operator** - This is responsible for the overall
+    functionality of the operator. This in turn makes use of other
+    components to establish the end goal of deciding whether a tuple is
+    a duplicate of some earlier tuple, or is a unique tuple.
+2.  **Managed State** - Since, all of the data cannot be stored in
+    memory, this component allows us to persist existing unique keys on
+    HDFS in form of buckets. This is also responsible for loading and
+    unloading of the buckets to and from the memory. This functions as
+    needed and loads the buckets requested by the Deduper.
+
+This was a very basic introduction to the functioning of the
+Deduper. Following sections will go into more detail on each of the
+components.
+
+Use cases - Basic Dedup
+====================================
+
+Dedup Key
+----------------------
+
+A dedup key is a set of one or more fields in the data tuple which
+acts as the key for the tuples.
+This is used by the deduper to compare tuples to arrive at the
+conclusion on whether two tuples are duplicates.
+
+
+
+Consider an example schema and two sample tuples
+
+`{Name, Phone, Email, Date, State, Zip, Country}`
+
+
+
+Tuple 1:
+```
+{
+  Austin U. Saunders,
+  +91-319-340-59385,
+  ausaunders@semperegestasurna.com,
+  2015-11-09 13:38:38,
+  Texas,
+  73301,
+  United States
+}
+```
+
+
+Tuple 2:
+```
+{
+  Austin U. Saunders,
+  +91-319-340-59385,
+  austin@semperegestasurna.com,
+  2015-11-09 13:39:38,
+  Texas,
+  73301,
+  United States
+}
+```
+
+
+Let us assume that the Dedup Key
+is `{Name, Phone}`. In
+this case, the two tuples are duplicates because the key fields are same
+in both the tuples. However, if we plan to make the Dedup Key as {Phone,
+Email}, then in this case, the two are unique tuples as the
+keys of both tuples do not match.
```

Comment:
However .... => However, if the Dedup Key is `{Phone, Email}`, the two are unique since the email values differ.